homeassistant/components/opensky/sensor.py

```diff
@@ -112,60 +155,128 @@ def state(self):
+                        route.get(ATTR_FLIGHT_NUMBER)
+                    )
+                else:
+                    flightnumber = ""
+            else:
+                route = None
+                flightnumber = ""
+            _LOGGER.debug("Flight: %s", flightnumber)
+
+            if flight in metadata and metadata[flight].get(ATTR_ICAO24) != "":
+                planemeta = self._get_plane_meta(metadata[flight].get(ATTR_ICAO24))
+            else:
+                planemeta = None
 
             data = {
+                ATTR_ICAO24: icao,
                 ATTR_CALLSIGN: flight,
+                "fligh": flightnumber,
                 ATTR_ALTITUDE: altitude,
                 ATTR_SENSOR: self._name,
             }
+            if route is not None:
+                # Merge all route info into data dictionary
+                data = {**data, **route}
+            if planemeta is not None:
+                data = {**data, **planemeta}
+            _LOGGER.debug("Fire event for: %s", str(data))
             self._hass.bus.fire(event, data)
 
     def update(self):
         """Update device state."""
         currently_tracked = set()
         flight_metadata = {}
-        states = self._session.get(OPENSKY_API_URL).json().get(ATTR_STATES)
-        for state in states:
-            flight = dict(zip(OPENSKY_API_FIELDS, state))
-            callsign = flight[ATTR_CALLSIGN].strip()
-            if callsign != "":
-                flight_metadata[callsign] = flight
-            else:
-                continue
-            missing_location = (
-                flight.get(ATTR_LONGITUDE) is None or flight.get(ATTR_LATITUDE) is None
-            )
-            if missing_location:
-                continue
-            if flight.get(ATTR_ON_GROUND):
-                continue
-            distance = util_location.distance(
-                self._latitude,
-                self._longitude,
-                flight.get(ATTR_LATITUDE),
-                flight.get(ATTR_LONGITUDE),
-            )
-            if distance is None or distance > self._radius:
-                continue
-            altitude = flight.get(ATTR_ALTITUDE)
-            if altitude > self._altitude and self._altitude != 0:
-                continue
-            currently_tracked.add(callsign)
-        if self._previously_tracked is not None:
-            entries = currently_tracked - self._previously_tracked
-            exits = self._previously_tracked - currently_tracked
-            self._handle_boundary(entries, EVENT_OPENSKY_ENTRY, flight_metadata)
-            self._handle_boundary(exits, EVENT_OPENSKY_EXIT, flight_metadata)
-        self._state = len(currently_tracked)
-        self._previously_tracked = currently_tracked
+        url_with_bbox = OPENSKY_API_URL % (
+            self._lat_min,
+            self._lon_min,
+            self._lat_max,
+            self._lon_max,
+        )
+        request = self._session.get(url_with_bbox)
```

Comment:
Please extract the device specific interface code and data parsing to a standalone library published on PyPI.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices

Reply:
> Please extract the device specific interface code and data parsing to a stand alone library published on PyPI.

OpenSky has an [official OpenSky Python API](https://opensky-network.org/apidoc/python.html#return-types) with a [repo on GitHub](https://github.com/openskynetwork/opensky-api) but not on PyPi.

On PyPi there are no less than four projects listed of OpenSky, two of which have been updated late last year. Maybe one of these projects could be used as the basis for the OpenSky integration instead of reinventing the wheel.

Reply:
We don't have a requirement that a new project is created on PyPI. If there's an existing maintained package, it's ok to use that.

Reply:
Yes, there is a Python API maintained by the opensky network core project (https://opensky-network.org/apidoc/python.html#). However, to what I could see, this API does not currently expose the plane specific information (e.g. plane type) and neither the route specific information (e.g. destination airport). These are the additional information that my changes add to the home assistant integration (https://github.com/home-assistant/home-assistant.io/pull/15775/files).
But for the information that was available prior to my changes this API might be a good replacement for the "manual" use of the REST API (https://opensky-network.org/apidoc/rest.html).

Reply:
The selected package needs to be published on PyPI.